<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Gota Morota" />

<meta name="date" content="2017-01-26" />

<title>Genome-wide allele frequencies</title>

<script src="day06_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="day06_files/bootstrap-3.3.6/css/bootstrap.min.css" rel="stylesheet" />
<script src="day06_files/bootstrap-3.3.6/js/bootstrap.min.js"></script>
<script src="day06_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="day06_files/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
<script src="day06_files/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
<link href="day06_files/readthedown-0.1/readthedown.css" rel="stylesheet" />
<script src="day06_files/readthedown-0.1/readthedown.js"></script>
<script src="day06_files/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; background-color: #dddddd; }
td.sourceCode { padding-left: 5px; }
code > span.kw { font-weight: bold; } /* Keyword */
code > span.dt { color: #800000; } /* DataType */
code > span.dv { color: #0000ff; } /* DecVal */
code > span.bn { color: #0000ff; } /* BaseN */
code > span.fl { color: #800080; } /* Float */
code > span.ch { color: #ff00ff; } /* Char */
code > span.st { color: #dd0000; } /* String */
code > span.co { color: #808080; font-style: italic; } /* Comment */
code > span.al { color: #00ff00; font-weight: bold; } /* Alert */
code > span.fu { color: #000080; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #ff0000; font-weight: bold; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #ff00ff; } /* SpecialChar */
code > span.vs { color: #dd0000; } /* VerbatimString */
code > span.ss { color: #dd0000; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #808080; font-style: italic; } /* Documentation */
code > span.an { color: #808080; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #808080; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #808080; font-weight: bold; font-style: italic; } /* Information */
</style>


</head>

<body>


<div id="content" data-toggle="wy-nav-shift">

<!-- tabsets -->
<script src="/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->

<nav id="nav-top" role="navigation" aria-label="top navigation">
    <a role="button" href="#" data-toggle="wy-nav-top"><span class="glyphicon glyphicon-menu-hamburger"></span></a>
</nav>


<div id="header">
<h1 class="title">Genome-wide allele frequencies</h1>
</div>


<div id="table-of-contents">
    <h2><a href="#content">Genome-wide allele frequencies</a></h2>
    <div id="text-table-of-contents">
      <ul>
      <li><a href="#overview">Overview</a></li>
      <li><a href="#read-a-file">Read a file</a></li>
      <li><a href="#compute-allele-frequencies-for-all-snps">Compute allele frequencies for all SNPs</a></li>
      <li><a href="#minor-allele-frequency">Minor allele frequency</a></li>
      <li><a href="#expectation-and-variance-of-allelic-counts">Expectation and variance of allelic counts</a></li>
      </ul>
    </div>
</div>

<div id="main">
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>We will learn how to compute 1) genome-wide allele frequencies and 2) expectation and variance of allelic counts.</p>
</div>
<div id="read-a-file" class="section level2">
<h2>Read a file</h2>
<p>Use the function <code>load</code> to reload the SNP matrix <code>W.Rda</code>, which we saved as an R object in the previous class.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="dt">file =</span> <span class="kw">file.choose</span>())</code></pre></div>
</div>
<div id="compute-allele-frequencies-for-all-snps" class="section level2">
<h2>Compute allele frequencies for all SNPs</h2>
<p>We have learned how to compute allele frequency of the first SNP using the <code>table</code> function. Here, we consider how to compute allele frequencies for the entire SNPs. Of course we can apply the <code>table</code> function manually one at a time. However, this approach takes too much time to compute allele frequencies for 5,000 SNPs. Recall that allele frequency of <em>A</em> is given by <span class="math display">\[ 
f(A) = p = \frac{2 \times (\text{no. of } AA \text{ individuals}) + 1 \times (\text{no. of } Aa \text{ individuals})}{2 \times \text{total no. of individuals}}.
\]</span> We can rewrite this equation into <span class="math display">\[ 
f(A) = p = \frac{(\text{no. of } A \text{ allele in the population})}{2 \times \text{total no. of individuals}}.
\]</span> This suggests that all we need is the number of <span class="math inline">\(A\)</span> allele or reference allele <span class="math inline">\(a\)</span> for each SNP. The <code>sum</code> function returns the number of reference allele <span class="math inline">\(A\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(W[, <span class="dv">1</span>])  <span class="co"># sum of A allele in the first SNP</span>
<span class="kw">sum</span>(W[, <span class="dv">2</span>])  <span class="co"># sum of A allele in the second SNP</span></code></pre></div>
<p>How to repeat this operation for 5,000 SNPs? The <code>colSums</code> function returns the sum of each column in a matrix as a vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colSums</span>(W)</code></pre></div>
<p>Note that <code>colSums(W)</code> gives the numerator of the above equation. We then divide this number by <span class="math inline">\(2 \times \text{total no. of individuals}\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">colSums</span>(W)/(<span class="dv">2</span> *<span class="st"> </span><span class="kw">nrow</span>(W))</code></pre></div>
<p>The variable <code>p</code> is a vector and it contains the allele frequencies of reference allele for 5,000 SNPs.</p>
<div id="exercise-1" class="section level3">
<h3>Exercise 1</h3>
<p>What is the allele frquency of reference allele in the 400th SNP?</p>
</div>
<div id="exercise-2" class="section level3">
<h3>Exercise 2</h3>
<p>What is the mean of reference allele frquencies in this population?</p>
</div>
</div>
<div id="minor-allele-frequency" class="section level2">
<h2>Minor allele frequency</h2>
<p>In most cases, people report a minor allele frequency, which is the frequency of the less frequent allele in a given SNP. We can convert allele frequencies into minor allele frquencies by using the <code>ifelse</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">maf &lt;-<span class="st"> </span><span class="kw">ifelse</span>(p &gt;<span class="st"> </span><span class="fl">0.5</span>, <span class="dv">1</span> -<span class="st"> </span>p, p)</code></pre></div>
<div id="exercise-3" class="section level3">
<h3>Exercise 3</h3>
<p>What is the minor allele frquency of reference allele in the 400th SNP?</p>
</div>
<div id="exercise-4" class="section level3">
<h3>Exercise 4</h3>
<p>What is the mean of minor allele frquencies?</p>
<p>Now letâ€™s visualize the minor allele frequencies for the first 500 SNPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(maf[<span class="dv">1</span>:<span class="dv">500</span>])</code></pre></div>
</div>
</div>
<div id="expectation-and-variance-of-allelic-counts" class="section level2">
<h2>Expectation and variance of allelic counts</h2>
<div id="exercise-5" class="section level3">
<h3>Exercise 5</h3>
<p>Recall that the expectation of genotype, <span class="math inline">\(E(W)\)</span>, is given by <span class="math inline">\(2p\)</span>, where <span class="math inline">\(p\)</span> is the frequency of reference allele. Verify that <span class="math inline">\(2p\)</span> is equal to the mean of each genotype obtained from the <code>colMeans()</code> function. Use the variables <code>W</code> and <code>p</code>.</p>
</div>
<div id="exercise-6" class="section level3">
<h3>Exercise 6</h3>
<p>Recall that the variance of genotype, <span class="math inline">\(Var(W)\)</span>, is given by <span class="math inline">\(2p(1-p)\)</span>. Verify that <span class="math inline">\(2p(1-p)\)</span> is close to the variance of each genotype obtained from the <code>var()</code> function. Use the variables <code>W</code> and <code>p</code>.</p>
</div>
</div>
</div>


</div>

<div id="postamble" data-toggle="wy-nav-shift" class="status">
<p class="author"><span class="glyphicon glyphicon-user"></span> Gota Morota</p>
<p class="date"><span class="glyphicon glyphicon-calendar"></span> January 26, 2017</p>
</div>


<script>
$(document).ready(function () {
 	  });
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
